addons/test_mail/tests/test_mail_mail.py

```diff
@@ -728,6 +733,86 @@ def test_mail_mail_send_server(self):
         self.assert_email_sent_smtp(message_from='user_1@test_2.com', emails_count=5, from_filter=self.server_domain.from_filter)
 
 
+    @mute_logger('odoo.addons.mail.models.mail_mail')
+    def test_mail_mail_send_server_attachment_to_download_link(self):
+        """ Test that when the mail size exceeds the max email size limit,
+        attachments are turned into download links added at the end of the
+        email content.
+
+        The test uses the low level method _send of mail_mail to perform the tests:
+        - using a specified server or the default one (to test command line parameters)
+        - sending one or more than one mail at once
+        - with mail that exceed (with one or more attachments) or not the limit
+        """
+        def count_attachments(message):
+            return (0 if isinstance(message, str)
+                    else sum(count_attachments(part) for part in message.get_payload()) if message.is_multipart()
+                    else 'attachment' in message.get('Content-Disposition', ''))
+
+        config_get = config.get
+        attachment_size_in_bytes = 1024 * 128
+        datas = base64.b64encode((''.join('.' for _ in range(0, attachment_size_in_bytes))).encode())
+
+        # max_email_size_inv produces the opposite of what is expected, it is assigned to the mail server not used
+        for idx_test, ((max_email_size, max_email_size_inv, turned_into_links, n_attachment), mail_server, n_mail) in \
+                enumerate(itertools.product(
+                    ((0.25, 0.1, False, 1),  # 1 attachment doesn't exceed max size
+                     (0.25, 1, True, 3),  # 3 attachment exceed max size
+                     (0.1, 0.25, True, 1),  # 1 attachment exceed max size
+                     ),
+                    (None, self.server_domain_2),  # mail server
+                    (1, 2)  # n_mail
+                )):
+            default_max_email_size = max_email_size if not mail_server else max_email_size_inv
+            self.server_domain_2.max_email_size = max_email_size if mail_server else max_email_size_inv
+            attachments = self.env['ir.attachment'].sudo().create([{
+                'name': f'attachment{idx_attachment}',
+                'res_name': 'test',
+                'res_model': self.test_record._name,
+                'res_id': self.test_record.id,
+                'datas': datas,
+            } for idx_attachment in range(n_attachment)])
+            mails = self.env['mail.mail'].create([{
+                'attachment_ids': attachments.ids,
+                'body_html': '<p>Test</p>',
+                'email_from': 'test@unknown_domain.com',
+                'email_to': f'user{idx_test}_{idx_mail}@example.com',
```

Comment:
If you really need unique recipients, maybe use test parameters so that we get rid our idx_test and enumerate ? to lessen a bit boilerplate :)